File: nonsense-poem-generator.ts
#!/usr/bin/env ts-node

/**
 * Nonsense Poem Generator
 * Generates random nonsensical poems using word banks and templates
 */

// Word banks for random selection
const adjectives = [
    'purple', 'bouncy', 'invisible', 'squeaky', 'jellied', 'grumpy',
    'wobbly', 'sparkly', 'fuzzy', 'peculiar', 'luminous', 'dizzy',
    'twisted', 'bubbly', 'slippery', 'crooked', 'jumbled', 'frazzled'
];

const nouns = [
    'penguin', 'toaster', 'umbrella', 'banana', 'telescope', 'accordion',
    'spatula', 'moonbeam', 'pickle', 'flamingo', 'waffle', 'bicycle',
    'teapot', 'mushroom', 'harmonica', 'walrus', 'pretzel', 'chandelier'
];

const verbs = [
    'dances', 'wiggles', 'somersaults', 'yodels', 'hiccups', 'juggles',
    'sneezes', 'bounces', 'twirls', 'prances', 'waddles', 'zigzags',
    'tumbles', 'flutters', 'wobbles', 'giggles', 'shuffles', 'leaps'
];

const adverbs = [
    'backwards', 'sideways', 'slowly', 'frantically', 'mysteriously', 'gleefully',
    'awkwardly', 'gracefully', 'clumsily', 'eagerly', 'suddenly', 'softly',
    'wildly', 'gently', 'rapidly', 'lazily', 'merrily', 'oddly'
];

const prepositions = [
    'under', 'over', 'through', 'around', 'beneath', 'beside',
    'above', 'below', 'between', 'among', 'beyond', 'inside'
];

const places = [
    'the garden', 'a cloud', 'the meadow', 'the ocean', 'a rainbow',
    'the attic', 'a dream', 'the forest', 'the stars', 'a puddle',
    'the moon', 'a teacup', 'the shadows', 'the wind', 'a bubble'
];

// Helper function to pick a random item from an array
function pick<T>(array: T[]): T {
    return array[Math.floor(Math.random() * array.length)];
}

// Poem templates
const poemTemplates = [
    // Template 1: Four-line AABB rhyme-ish structure
    () => {
        const adj1 = pick(adjectives);
        const noun1 = pick(nouns);
        const verb1 = pick(verbs);
        const adv1 = pick(adverbs);
        
        const adj2 = pick(adjectives);
        const noun2 = pick(nouns);
        const prep1 = pick(prepositions);
        const place1 = pick(places);
        
        return `The ${adj1} ${noun1} ${verb1} ${adv1},
While ${adj2} ${noun2}s rest ${prep1} ${place1}.
With ${pick(adjectives)} delight and ${pick(adjectives)} glee,
They ${pick(verbs)} together wild and free!`;
    },
    
    // Template 2: Haiku-style (5-7-5 syllables approximation)
    () => {
        return `${pick(adjectives)} ${pick(nouns)}
${pick(verbs)} ${pick(adverbs)} ${pick(prepositions)} ${pick(places)}
${pick(adjectives)} ${pick(nouns)} ${pick(verbs)}`;
    },
    
    // Template 3: Nonsense narrative
    () => {
        const noun1 = pick(nouns);
        const noun2 = pick(nouns);
        const noun3 = pick(nouns);
        
        return `Once a ${pick(adjectives)} ${noun1} met a ${pick(adjectives)} ${noun2},
They ${pick(verbs)} ${pick(adverbs)} ${pick(prepositions)} ${pick(places)}.
"${pick(adjectives).charAt(0).toUpperCase() + pick(adjectives).slice(1)} ${pick(nouns)}!" the ${noun1} cried,
As a ${pick(adjectives)} ${noun3} ${pick(verbs)} by its side.`;
    },
    
    // Template 4: Repetitive structure
    () => {
        const verb = pick(verbs);
        const adv = pick(adverbs);
        const adj = pick(adjectives);
        const noun = pick(nouns);
        
        return `The ${noun} ${verb},
${adv} it ${verb},
${adj} and ${pick(adjectives)} it ${verb},
Oh, how that ${noun} ${verb}!`;
    },
    
    // Template 5: Question and answer
    () => {
        const noun1 = pick(nouns);
        const noun2 = pick(nouns);
        
        return `What happens when a ${pick(adjectives)} ${noun1}
Meets a ${pick(adjectives)} ${noun2} ${pick(prepositions)} ${pick(places)}?
They ${pick(verbs)} and ${pick(verbs)} ${pick(adverbs)},
Creating ${pick(adjectives)} surprises!`;
    },
    
    // Template 6: List poem
    () => {
        return `I saw:
A ${pick(adjectives)} ${pick(nouns)} that ${pick(verbs)} ${pick(adverbs)},
A ${pick(adjectives)} ${pick(nouns)} ${pick(prepositions)} ${pick(places)},
And a ${pick(adjectives)} ${pick(nouns)} doing ${pick(adjectives)} things,
All while the ${pick(adjectives)} ${pick(nouns)} ${pick(verbs)}!`;
    }
];

// Generate a single poem
function generatePoem(): string {
    const template = pick(poemTemplates);
    return template();
}

// Generate multiple poems
function generatePoems(count: number): void {
    console.log('='.repeat(60));
    console.log('  🎭 NONSENSE POEM GENERATOR 🎭');
    console.log('='.repeat(60));
    console.log();
    
    for (let i = 1; i <= count; i++) {
        console.log(`Poem #${i}:`);
        console.log('-'.repeat(60));
        console.log(generatePoem());
        console.log();
        
        if (i < count) {
            console.log();
        }
    }
    
    console.log('='.repeat(60));
}

// Main execution
if (require.main === module) {
    const args = process.argv.slice(2);
    const count = args.length > 0 ? parseInt(args[0], 10) : 3;
    
    if (isNaN(count) || count < 1) {
        console.error('Usage: ts-node nonsense-poem-generator.ts [number_of_poems]');
        console.error('Example: ts-node nonsense-poem-generator.ts 5');
        process.exit(1);
    }
    
    generatePoems(count);
}

// Export for use as a module
export { generatePoem, generatePoems };

File: nonsense-poem-generator.pl
% Nonsense Poem Generator - Prolog Edition
% To run: swipl -g main -t halt nonsense-poem-generator.pl [count]

:- initialization(main, main).

% Word banks
adjective('purple').
adjective('bouncy').
adjective('invisible').
adjective('squeaky').
adjective('jellied').
adjective('grumpy').
adjective('wobbly').
adjective('sparkly').
adjective('fuzzy').
adjective('peculiar').
adjective('luminous').
adjective('dizzy').

noun('penguin').
noun('toaster').
noun('umbrella').
noun('banana').
noun('telescope').
noun('accordion').
noun('spatula').
noun('moonbeam').
noun('pickle').
noun('flamingo').
noun('waffle').
noun('bicycle').

verb('dances').
verb('wiggles').
verb('somersaults').
verb('yodels').
verb('hiccups').
verb('juggles').
verb('sneezes').
verb('bounces').
verb('twirls').
verb('prances').
verb('waddles').
verb('zigzags').

adverb('backwards').
adverb('sideways').
adverb('slowly').
adverb('frantically').
adverb('mysteriously').
adverb('gleefully').
adverb('awkwardly').
adverb('gracefully').
adverb('clumsily').
adverb('eagerly').

preposition('under').
preposition('over').
preposition('through').
preposition('around').
preposition('beneath').
preposition('beside').
preposition('above').
preposition('below').
preposition('between').
preposition('among').

place('the garden').
place('a cloud').
place('the meadow').
place('the ocean').
place('a rainbow').
place('the attic').
place('a dream').
place('the forest').
place('the stars').
place('a puddle').

% Pick random element from a category
pick_random(Category, Word) :-
    findall(W, call(Category, W), Words),
    length(Words, Len),
    Len > 0,
    random_between(1, Len, Index),
    nth1(Index, Words, Word).

% Capitalize first letter
capitalize_first(Word, Capitalized) :-
    atom_chars(Word, [First|Rest]),
    upcase_atom(First, Upper),
    atom_chars(Capitalized, [Upper|Rest]).

% Poem templates
generate_template1 :-
    pick_random(adjective, Adj1),
    pick_random(noun, Noun1),
    pick_random(verb, Verb1),
    pick_random(adverb, Adv1),
    pick_random(adjective, Adj2),
    pick_random(noun, Noun2),
    pick_random(preposition, Prep1),
    pick_random(place, Place1),
    pick_random(adjective, Adj3),
    pick_random(adjective, Adj4),
    pick_random(verb, Verb2),
    format('The ~w ~w ~w ~w,~n', [Adj1, Noun1, Verb1, Adv1]),
    format('While ~w ~ws rest ~w ~w.~n', [Adj2, Noun2, Prep1, Place1]),
    format('With ~w delight and ~w glee,~n', [Adj3, Adj4]),
    format('They ~w together wild and free!~n', [Verb2]).

generate_template2 :-
    pick_random(verb, Verb),
    pick_random(noun, Noun),
    pick_random(adjective, Adj1),
    pick_random(adjective, Adj2),
    pick_random(adverb, Adv),
    format('The ~w ~w,~n', [Noun, Verb]),
    format('~w it ~w,~n', [Adv, Verb]),
    format('~w and ~w it ~w,~n', [Adj1, Adj2, Verb]),
    format('Oh, how that ~w ~w!~n', [Noun, Verb]).

generate_template3 :-
    pick_random(adjective, Adj1),
    pick_random(noun, Noun1),
    pick_random(verb, Verb1),
    pick_random(adverb, Adv1),
    pick_random(adjective, Adj2),
    pick_random(noun, Noun2),
    pick_random(preposition, Prep1),
    pick_random(place, Place1),
    pick_random(adjective, Adj3),
    pick_random(noun, Noun3),
    pick_random(verb, Verb2),
    format('I saw:~n', []),
    format('A ~w ~w that ~w ~w,~n', [Adj1, Noun1, Verb1, Adv1]),
    format('A ~w ~w ~w ~w,~n', [Adj2, Noun2, Prep1, Place1]),
    format('And a ~w ~w doing ~w things,~n', [Adj3, Noun3, Adj3]),
    format('All while the ~w ~w ~w!~n', [Adj1, Noun2, Verb2]).

% Generate a single poem
generate_poem :-
    random_between(1, 3, Template),
    (   Template = 1 -> generate_template1
    ;   Template = 2 -> generate_template2
    ;   Template = 3 -> generate_template3
    ).

% Generate multiple poems
generate_poems(0) :- !.
generate_poems(N) :-
    N > 0,
    Count is 4 - N,
    format('Poem #~w:~n', [Count]),
    format('------------------------------------------------------------~n', []),
    generate_poem,
    nl,
    (N > 1 -> nl ; true),
    N1 is N - 1,
    generate_poems(N1).

% Main entry point
main(Argv) :-
    (   Argv = [CountAtom|_],
        atom_number(CountAtom, Count),
        Count > 0
    ->  true
    ;   Count = 3
    ),
    format('============================================================~n', []),
    format('  🎭 NONSENSE POEM GENERATOR (Prolog) 🎭~n', []),
    format('============================================================~n', []),
    nl,
    generate_poems(Count),
    format('============================================================~n', []).

main :-
    current_prolog_flag(argv, Argv),
    main(Argv).
